feat(subscriptions): Implement SubscriptionScheduler

wedamija opened the pull request:
The `SubscriptionScheduler` keeps a dict of `Subscription` objects that it manages. When `find` is
passed an `Interval[datetime]`, return all `ScheduledTask[Subscription]` objects that should be
run within that interval.

wedamija committed:
feat(subscriptions): Implement SubscriptionScheduler

The `SubscriptionScheduler` keeps a dict of `Subscription` objects that it manages. When `find` is
passed an `Interval[datetime]`, return all `ScheduledTask[Subscription]` objects that should be
run within that interval.

tkaemming commented on snuba/subscriptions/scheduler.py:
```diff
@@ -39,3 +42,47 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
+    def clear(self) -> None:
+        # XXX: Not part of the standard interface. Since we want to periodically clear
+        # the cache, whatever is managing this should probably clear and refetch
+        self.__subscriptions = {}
```
Should these be internal methods? I kind of assumed the constructor would just be passed the `RedisSubscriptionStore` and any sort of caching of that state would happen privately in the implementation, curious if you have a different approach in mind.

wedamija replied:
I based these extra methods off of https://github.com/getsentry/snuba/compare/scheduler#diff-22926dd5eda88654f5e7bb17ae6165aaR8-R26, and assumed that something external would be passing updates in (like the `SubscriptionConsumer` or whatever we end up calling it).

I don't mind too much if this class just gets passed a datastore and receives updates from it though.

wedamija replied:
Hmm, passing in the datastore doesn't work so well after your refactor, since we expect this to return `Iterator[ScheduledTask[Subscription]]`, and the datastore can only really return `SubscriptionData` rows. I suppose that if we pass the partition id to this class then we can rebuild the `SubscriptionIdentifier` with that + the UUID. Thoughts?

tkaemming replied:
> I suppose that if we pass the partition id to this class then we can rebuild the SubscriptionIdentifier with that + the UUID. Thoughts?

Yeah, I think that is fine. I don't love that both the store and the scheduler will have the partition ID as attributes, but that should be able to be changed reasonably easily later if it's a problem. The only other options that I can see would be to either:

1. expose the scheduler interface on `RedisSubscriptionDataStore`, or
2. make the subscription store deal with `SubscriptionIdentifier` keys instead of `UUID` ones (this one doesn't seem so bad, to be honest)

wedamija replied:
2. Doesn't seem so bad... Slight redundancy in the keys, but that's all. Kind of weird though - if we did that, then we may as well just pass the `Subscription` object when creating?

wedamija replied:
Gonna just go with passing the partition id for now. We can change it as needed (and after your other pr lands)

tkaemming replied:
Yeah, I think ideally the input data structure matches the output one, so `Subscription` objects would go in, and `Subscription` objects would come out.

That would make it so there would be a single interface for dealing with all of the partitions at once on the write path, which is kind of nice — we could just set it in once in the constructor of the `Creator` and not have to instantiate it all of the time.

It is a bit problematic for the worker, which only needs the subscriptions from a single partition when scheduling ([example here](https://github.com/getsentry/snuba/blob/388a89cd7507bd9667425229fbc42859e42ab5c2/snuba/subscriptions/worker.py#L36-L45), this commit may be GC'ed in the future due to rebasing). Ideally we wouldn't have to filter out results from `store.all` or `scheduler.find` based on their partition, so we might need to put a bit more consideration into how these interact.

tkaemming commented on snuba/subscriptions/scheduler.py:
```diff
@@ -39,3 +42,47 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
+            ):
+                yield ScheduledTask(
+                    datetime.fromtimestamp(i * resolution), subscription,
+                )
```
I forget why we decided this (I thought there was a specific reason), but I don't think this holds true wit this implementation: https://github.com/getsentry/snuba/blob/cf3f35cb050d50a513fe2e6f85be6f32de24f40c/snuba/subscriptions/scheduler.py#L38-L39

wedamija replied:
I think the main thing is that we need to return things ordered by timestamp within a subscription. It doesn't matter much other than that imo. I might just update the docstring to include that?

Alternatively we can just switch the loops

tkaemming replied:
No strong opinion about this one, as long as the implementation and the docstring match I'm fine with updating whichever one is easier.

wedamija replied:
I ended up switching things around. It actually makes the code a little cleaner, and means we don't have to iterate through every subscription on every tick.

wedamija committed:
- Pass `SubscriptionStore` to `SubscriptionScheduler` and have it manage fetching/caching items by
itself
- Remove set/delete/etc from `SubscriptionScheduler`
- Ensure that `SubscriptionScheduler.find` returns items in timestamp order. This also improves
  efficiency a little in that we only scan all subscriptions every second rather than every tick.

wedamija commented:
Note that this will likely change further after https://github.com/getsentry/snuba/pull/706 merges

wedamija committed:
Merge branch 'master' into danf/scheduler

wedamija committed:
Modify implementation to match refactored classes

wedamija committed:
mypy fixes

tkaemming commented on snuba/datasets/discover.py:
```diff
@@ -287,10 +287,7 @@ def get_query_processors(self) -> Sequence[QueryProcessor]:
-                            "_tags_flattened",
-                            {"timestamp"},
-                            BEGINNING_OF_TIME,
+                            "tags", "_tags_flattened", {"timestamp"}, BEGINNING_OF_TIME,
```
argh

tkaemming commented on snuba/subscriptions/store.py:
```diff
@@ -8,7 +9,32 @@
-class RedisSubscriptionDataStore:
+class SubscriptionDataStore(abc.ABC):
+    @abc.abstractmethod
+    def create(self, subscription_id: UUID, subscription: SubscriptionData) -> None:
```
Kind of surprised and disappointed that mypy didn't catch this — the types here and in `delete` match `RedisSubscriptionDataStore`, but the argument names don't. I changed this in https://github.com/getsentry/snuba/pull/706#discussion_r368192750 so that it'd be easier to pull out into a generic later as a mapping that handled encoding/decoding values, which seems useful. We don't have a clear use case for it at the moment, so I don't really mind which approach we take, but it should be consistent. (Personally, I think `subscription_id` is a bit overloaded with `SubscriptionIdentifier` but I guess you could make a reasonable argument that `key` is overly generic as well.)

tkaemming commented on snuba/subscriptions/scheduler.py:
```diff
@@ -39,3 +42,42 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
+            for subscription in subscriptions:
+                resolution = int(subscription.data.resolution.total_seconds())
+                if i % resolution == 0:
+                    yield ScheduledTask(datetime.fromtimestamp(i), subscription)
```
I wish that we would have called the `resolution` concept something like `frequency` or `period` instead. Resolution sounds too much like the time window to me.

I think this approach is fine and will probably be reasonably easy to optimize later, since we will probably need to time shift these values to avoid bursts on common time alignments. At that point we can store something like:

```python
self.__subscriptions = {
  60: {
    0: [],  # all items to be executed when timestamp % resolution == 0
    …,
    59: [],
  },
  300: {…},
}
```

However we determine that the remainder value should result in items that can be efficiently coalesced together falling into the same slot. I don't think we should worry about that quite yet, but it will probably happen pretty soon. I guess also be aware of this for the UI side — schedules will not always cleanly align on convenient time intervals.

I also think that if we are going to go this route (discretizing time to integers), we should make sure that subscriptions do not have a fractional component in `resolution` when they are instantiated.

wedamija replied:
I don't mind if we want to rename `resolution` on the snuba side, it's pretty low effort.

tkaemming commented on snuba/subscriptions/scheduler.py:
```diff
@@ -39,3 +42,42 @@ def find(self, interval: Interval[datetime]) -> Iterator[ScheduledTask[TTask]]:
+        self, interval: Interval[datetime]
+    ) -> Iterator[ScheduledTask[Subscription]]:
+        subscriptions = self.__get_subscriptions(interval.lower)
+        for i in range(
```
```suggestion
        for timestamp in range(
```

tkaemming reviewed (approved):
A couple minor things inline.

wedamija committed:
Merge branch 'master' into danf/scheduler

wedamija committed:
Code review feedback
